File: Data Structures & Algorithms/max-area-of-island/submission-3.java
class Solution {
    int[][] dirs = {{-1,0}, {1,0}, {0, -1}, {0,1}};

    public int maxAreaOfIsland(int[][] grid) {
        int res = 0;
        for(int r=0; r<grid.length; r++){
            for(int c = 0; c<grid[0].length; c++){
                if(grid[r][c] == 1){
                    res = Math.max(res, dfs(grid, r, c));
                }
            }
        }

        return res;
    }

    int dfs(int[][] grid, int r, int c){
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length 
            || grid[r][c] == 0){
                return 0;
        }

        grid[r][c] = 0;
        int res = 1;
        for(int[] dir:dirs){
                res += dfs(grid, r + dir[0], c + dir[1]);
        }

        return res;
    }
}

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-3.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        rows = len(matrix)
        cols = len(matrix[0])

        dirs = [[-1,0], [1, 0], [0, -1], [0, 1]]

        def dfs(i, j, pre):
            if i < 0 or j < 0 or i == rows or j == cols or pre >= matrix[i][j]:
                return 0

            res = 0
            for dir in dirs:
                res = max(res, 1 + dfs(i+dir[0], j + dir[1], matrix[i][j]))

            return res

        res = 0
        for i in range(rows):
            for j in range(cols):
                res = max(res, dfs(i, j, -1))

        return res
